packages/dd-trace/src/plugins/util/web.js

```diff
@@ -502,6 +569,73 @@ function getProtocol (req) {
   return 'http'
 }
 
+const v4PrivateCidrs = [
+  '127.0.0.0/8',
+  '10.0.0.0/8',
+  '172.16.0.0/12',
+  '192.168.0.0/16',
+  '169.254.0.0/16'
+].map(cidr => ip6addr.createCIDR(cidr)).map(cidr => [ cidr.address().toLong(), cidr.broadcast().toLong() ])
+
+const v6PrivateCidrs = [
+  '::1/128',
+  'fec0::/10',
+  'fe80::/10',
+  'fc00::/7',
+  'fd00::/8'
+].map(cidr => ip6addr.createCIDR(cidr))
+
+function findFirstIp (str) {
+  let firstPrivateIp
+  const splitted = str.split(',')
+
+  for (let i = 0; i < splitted.length; i++) {
+    const chunk = splitted[i].trim()
+
+    // TODO: strip port and interface data ?
+
+    let ip
+
+    try {
+      ip = ip6addr.parse(chunk)
+    } catch {
+      // not a valid ip, skip to next chunk
+      continue
+    }
+
+    let isPrivate = false
+
+    if (ip._attrs.ipv4Mapped || ip._attrs.ipv4Bare) {
+      // optimization to compare ipv4s numerically
+      const numericIp = ((ip._fields[6] << 16) >>> 0) + ip._fields[7]
+
+      for (let j = 0; j < v4PrivateCidrs.length; j++) {
+        if (numericIp >= v4PrivateCidrs[j][0] && numericIp <= v4PrivateCidrs[j][1]) {
+          isPrivate = true
+          break
+        }
+      }
+    } else {
+      for (let j = 0; j < v6PrivateCidrs.length; j++) {
+        if (v6PrivateCidrs[j].contains(ip)) {
+          isPrivate = true
+          break
+        }
+      }
+    }
+
+    if (isPrivate) {
```

Comment:
Why are these handled differently?

Reply:
Because in case of a mix of public and private IPs, we want to select public IPs in priority

Reply:
But then we're arbitrarily picking a specific header instead of being consistent, which is confusing and would almost definitely be incorrect.

Reply:
It's also worth noting that the above code could be redone more efficiently in any case by minimizing the number of passes, using sets, etc.